bug: rendering issue with ion-segment on iOS

### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

v7.x

### Current Behavior

After updating my app to "Ionic 7", I encountered a bug in "ion-segment": say you have a page with an `ion-segment` with 5 segments and the third is pre-selected through the `[(ngModel)]` or `[value]` binding. In addition to the value bindings `[scrollable]="true"` is also set (scrollable is actually used in my real application). Then if you navigate into this page it takes a moment to render the `ion-segment` component at the correct position.

As far as I can tell this bug is only reproducible on iOS (Simulator or Device). It's reproducible in Safari on my desktop computer but not in Chrome.

I've made a short clip to see the bug in action and also made the code to reproduce it available.


https://github.com/ionic-team/ionic-framework/assets/128096/70530696-95c2-4750-8e2e-318251e114d3



https://github.com/trytuna/IonicSegmentBug

### Expected Behavior

Should render without any visual glitches

### Steps to Reproduce

1. git clone https://github.com/trytuna/IonicSegmentBug
2. cd IonicSegmentBug
3. npm install
4. ionic build
5. npx cap sync ios
6. npx cap open ios
7. run it on an iPhone (e.g. 15 Pro iOS 17.0)
8. Navigate into the page with the bug by clicking on the blue "Bug"-Button

### Code Reproduction URL

https://github.com/trytuna/IonicSegmentBug

### Ionic Info

% ionic info
[WARN] Error loading @capacitor/android package.json: Error: Cannot find module '@capacitor/android/package.json'
       
       Require stack:
       - /Users/timo/.nvm/versions/node/v16.17.1/lib/node_modules/@ionic/cli/lib/project/index.js
       - /Users/timo/.nvm/versions/node/v16.17.1/lib/node_modules/@ionic/cli/lib/index.js
       - /Users/timo/.nvm/versions/node/v16.17.1/lib/node_modules/@ionic/cli/index.js
       - /Users/timo/.nvm/versions/node/v16.17.1/lib/node_modules/@ionic/cli/bin/ionic

Ionic:

   Ionic CLI                     : 7.1.1 (/Users/timo/.nvm/versions/node/v16.17.1/lib/node_modules/@ionic/cli)
   Ionic Framework               : @ionic/angular 7.5.0
   @angular-devkit/build-angular : 16.2.6
   @angular-devkit/schematics    : 16.2.6
   @angular/cli                  : 16.2.6
   @ionic/angular-toolkit        : 9.0.0

Capacitor:

   Capacitor CLI      : 5.5.0
   @capacitor/android : not installed
   @capacitor/core    : 5.5.0
   @capacitor/ios     : 5.5.0

Utility:

   cordova-res : not installed globally
   native-run  : 1.7.3

System:

   NodeJS : v16.17.1 (/Users/timo/.nvm/versions/node/v16.17.1/bin/node)
   npm    : 9.8.0
   OS     : macOS Unknown


### Additional Information

_No response_

--- a/core/src/components/segment/segment.tsx
+++ b/core/src/components/segment/segment.tsx
@@ -145,7 +145,13 @@ export class Segment implements ComponentInterface {
      * before we can scroll.
      */
     raf(() => {
-      this.scrollActiveButtonIntoView();
+     /**
+      * When the segment loads for the first
+      * time we just want to snap the active button into
+      * place instead of scroll. Smooth scrolling should only
+      * happen when the user interacts with the segment.
+      */
+      this.scrollActiveButtonIntoView(false);
     });
 
     this.gesture = (await import('../../utils/gesture')).createGesture({
@@ -305,31 +311,53 @@ export class Segment implements ComponentInterface {
     }
   }
 
-  private scrollActiveButtonIntoView() {
-    const { scrollable, value } = this;
+  private scrollActiveButtonIntoView(smoothScroll = true) {
+    const { scrollable, value, el } = this;
 
     if (scrollable) {
       const buttons = this.getButtons();
       const activeButton = buttons.find((button) => button.value === value);
       if (activeButton !== undefined) {
-        /**
-         * Scrollable segment buttons should be
-         * centered within the view including
-         * buttons that are partially offscreen.
-         */
-        activeButton.scrollIntoView({
-          behavior: 'smooth',
-          inline: 'center',
+        const scrollContainerBox = el.getBoundingClientRect();
+        const activeButtonBox = activeButton.getBoundingClientRect();
 
-          /**
-           * Segment should scroll on the
-           * horizontal axis. `block: 'nearest'`
-           * ensures that the vertical axis
-           * does not scroll if the segment
-           * as a whole is already in view.
-           */
-          block: 'nearest',
-        });
+        /**
+         * Subtract the active button x position from the scroll
+         * container x position. This will give us the x position
+         * of the active button within the scroll container.
+         */
+        const activeButtonLeft = activeButtonBox.x - scrollContainerBox.x;
+
+        /**
+         * If we just used activeButtonLeft, then the active button
+         * would be aligned with the left edge of the scroll container.
+         * Instead, we want the segment button to be centered. As a result,
+         * we subtract half of the scroll container width. This will position
+         * the left edge of the active button at the midpoint of the scroll container.
+         * We then add half of the active button width. This will position the active
+         * button such that the midpoint of the active button is at the midpoint of the
+         * scroll container.
+         */
+        const centeredX = activeButtonLeft - (scrollContainerBox.width / 2) + (activeButtonBox.width / 2);
+
+        /**
+         * We intentionally use scrollBy here instead of scrollIntoView
+         * to avoid a WebKit bug where accelerated animations break
+         * when using scrollIntoView. Using scrollIntoView will cause the
+         * segment container to jump during the transition and then snap into place.
+         * This is because scrollIntoView can potentially cause parent element
+         * containers to also scroll. scrollBy does not have this same behavior, so
+         * we use this API instead.
+         *
+         * Note that if there is not enough scrolling space to center the element
+         * within the scroll container, the browser will attempt
+         * to center by as much as it can.
+         */
+        this.el.scrollBy({
+          top: 0,
+          left: centeredX,
+          behavior: smoothScroll ? 'smooth' : 'instant'
+        })
       }
     }
   }

--- a/core/src/components/segment/segment.tsx
+++ b/core/src/components/segment/segment.tsx
@@ -353,7 +353,7 @@ export class Segment implements ComponentInterface {
          * within the scroll container, the browser will attempt
          * to center by as much as it can.
          */
-        this.el.scrollBy({
+        el.scrollBy({
           top: 0,
           left: centeredX,
           behavior: smoothScroll ? 'smooth' : 'instant'

--- a/core/src/components/segment/segment.tsx
+++ b/core/src/components/segment/segment.tsx
@@ -145,12 +145,12 @@ export class Segment implements ComponentInterface {
      * before we can scroll.
      */
     raf(() => {
-     /**
-      * When the segment loads for the first
-      * time we just want to snap the active button into
-      * place instead of scroll. Smooth scrolling should only
-      * happen when the user interacts with the segment.
-      */
+      /**
+       * When the segment loads for the first
+       * time we just want to snap the active button into
+       * place instead of scroll. Smooth scrolling should only
+       * happen when the user interacts with the segment.
+       */
       this.scrollActiveButtonIntoView(false);
     });
 
@@ -338,7 +338,7 @@ export class Segment implements ComponentInterface {
          * button such that the midpoint of the active button is at the midpoint of the
          * scroll container.
          */
-        const centeredX = activeButtonLeft - (scrollContainerBox.width / 2) + (activeButtonBox.width / 2);
+        const centeredX = activeButtonLeft - scrollContainerBox.width / 2 + activeButtonBox.width / 2;
 
         /**
          * We intentionally use scrollBy here instead of scrollIntoView
@@ -356,8 +356,8 @@ export class Segment implements ComponentInterface {
         el.scrollBy({
           top: 0,
           left: centeredX,
-          behavior: smoothScroll ? 'smooth' : 'instant'
-        })
+          behavior: smoothScroll ? 'smooth' : 'instant',
+        });
       }
     }
   }

